pkg/cmd/token.go

```diff
@@ -0,0 +1,152 @@
+package cmd
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/puppetlabs/relay/pkg/client/openapi"
+	"github.com/puppetlabs/relay/pkg/config"
+	"github.com/puppetlabs/relay/pkg/debug"
+	"github.com/spf13/cobra"
+)
+
+func newTokensCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "tokens",
+		Short: "Manage API tokens",
+		Args:  cobra.MinimumNArgs(1),
+	}
+
+	cmd.AddCommand(newCreateToken())
+	cmd.AddCommand(newListTokens())
+	cmd.AddCommand(newRevokeToken())
+
+	return cmd
+}
+
+func newCreateToken() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "create [token name]",
+		Short: "Create API token",
+		Args:  cobra.MinimumNArgs(1),
+		RunE:  doCreateToken,
+	}
+
+	cmd.Flags().BoolP("use", "u", true, "Configure the CLI to use the generated API token")
+	cmd.Flags().StringP("file", "f", "", "Write the generated token to the supplied file")
+
+	return cmd
+}
+
+func newRevokeToken() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "revoke [token id]",
+		Short: "Revoke API token",
+		Args:  cobra.MinimumNArgs(1),
+		RunE:  doRevokeToken,
+	}
+
+	return cmd
+}
+
+func newListTokens() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "List API tokens",
+		Args:  cobra.MaximumNArgs(1),
+		RunE:  doListTokens,
+	}
+
+	cmd.Flags().BoolP("all", "a", false, "Show all account tokens")
+
+	return cmd
+}
+
+func doCreateToken(cmd *cobra.Command, args []string) error {
```

Comment:
A couple of thoughts on this one...

* It looks like if you specify `--use=false` without `--file` you won't get to see the token at all. I think we should either make this an error (i.e., make those options mutually exclusive) or actually just print the token to the screen in this case.
* For `--use=true` (the default) I think we should probably stick a little blurb at the end letting the user know we updated the authentication mechanism for them. E.g., "The generated token has been added to the cached credentials for this Relay CLI installation. To create a token without using it for authentication locally, specify `--use=false`. To clear your cached credentials, use `relay config auth clear`.")
* For `--use=false` without `--file`, if we decide to print the token to the screen, we should let the user know how they can auth with it. E.g., "To use this token for authentication with the Relay CLI, run `echo '<token>' | relay auth login --stdin`."
* For any use of `--file`, we should let the user know how they can use the file too. E.g., "Your token was written to \<file>. You can use this file to authenticate to the Relay CLI by running `relay auth login --file=<file>`."

Reply:
Will do. Good suggestions, thanks.

It originally printed out to the screen if not using the file option, but I've gone back and forth on whether that's appropriate (i.e. accidental being logged in external systems). But I've already "misplaced" one secret that way...